lib/deaf_grandma.rb

```diff
@@ -19,13 +23,35 @@ def run!
 
 
   def speak(input)
+    if bye? input then
+      @bye_counter += 1
+    end
 
-    #Implement your code here <<<<<<<<<
+    if @bye_counter >= @bye_limit then
+      return @farewell
+    end
 
+    if can_hear? input then
+      return @heard_response
+    else
+      return @not_heard_response
+    end
   end
 
   private
 
+  def bye?(said)
+    said == "BYE"
+  end
+
+  def can_hear?(said)
+    said.chars.all? { |character| upcase? character }
+  end
+
+  def upcase?(char)
+    char == char.upcase
+  end
```

Comment:
`upcase` is a string method, so you can just check `said == said.upcase` instead of checking each character individually.